alicloud/resource_alicloud_cs_kubernetes_autoscaler.go

```diff
@@ -0,0 +1,552 @@
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(90 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"nodepools": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"taints": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"labels": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+				MaxItems: 30,
+			},
+			"utilization": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"cool_down_duration": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"defer_scale_in_duration": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+// resourceAlicloudCSKubernetesAutoscalerCreate define how to create autoscaler
+func resourceAlicloudCSKubernetesAutoscalerCreate(d *schema.ResourceData, meta interface{}) error {
+
+	var clusterId string
+	if value, ok := d.GetOk("cluster_id"); ok == true {
+		clusterId = value.(string)
+	} else {
+		return WrapError(fmt.Errorf("clusterId must be provided instead of %s", "nil"))
+	}
+
+	err := resourceAlicloudCSKubernetesAutoscalerUpdate(d, meta)
+
+	if err != nil {
+		return WrapError(err)
+	}
+	// set unique id of tf state
+	d.SetId(fmt.Sprintf("%s:%s", clusterId, clusterAutoscaler))
+	return nil
+}
+
+// resourceAlicloudCSKubernetesAutoscalerRead no need to implement
+func resourceAlicloudCSKubernetesAutoscalerRead(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+// resourceAlicloudCSKubernetesAutoscalerUpdate define how to update autoscaler configuration
+func resourceAlicloudCSKubernetesAutoscalerUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*connectivity.AliyunClient)
+	csService := CsService{client}
+
+	// any changes need to ready autoscaler.
+	if d.HasChange("nodepools") || d.HasChange("utilization") || d.HasChange("cool_down_duration") || d.HasChange("defer_scale_in_duration") {
+
+		regionId := client.RegionId
+
+		var clusterId, utilization, coolDownDuration, deferScaleInDuration string
+		var ok bool
+		var value interface{}
+
+		if value, ok = d.GetOk("cluster_id"); ok == true {
+			clusterId = value.(string)
+		} else {
+			return WrapError(fmt.Errorf("please provide your cluster id in region %s", regionId))
+		}
+
+		// check utilization value
+		if value, ok = d.GetOk("utilization"); ok == true {
+			utilization = value.(string)
+		} else {
+			return WrapError(fmt.Errorf("please provide utilization of cluster %s", clusterId))
+		}
+
```

Comment:
Because the `utilization` is required, the line 148-152 can be instead by `utilization := d.Get("utilization").(string)`